# Урок 3
Описательная статистика. Качественные и количественные характеристики популяции. Графическое представление данных
Домашнее задание

Задача 1

Даны значения зарплат из выборки выпускников:

100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 230, 24, 57, 55, 70, 75, 65, 84, 90, 150

Используя только встроенные питоновские функции и структуры данных (т.е. без библиотек numpy, pandas и др.) посчитать среднее арифметическое, смещённую и несмещённую оценки дисперсии, среднее квадратичное отклонение для данной выборки.

Можно затем посчитать те же значения с использованием библиотек, чтобы проверить себя.

In [1]:
wages = [100, 80, 75, 77, 89, 33, 45, 25, 65, 17, 30, 230, 24, 57, 55, 70, 75, 65, 84, 90, 150]

In [2]:
# среднее арифметическое
mean_wages = sum(wages)/len(wages)
print(mean_wages)

73.14285714285714


In [3]:
# смещённую и несмещённую оценки дисперсии
wages_els_mmean_sqr_sum = 0
for i in range(len(wages)):
    el_mmean_sqr = (wages[i] - mean_wages)**2
    wages_els_mmean_sqr_sum += el_mmean_sqr
print(wages_els_mmean_sqr_sum)

44836.571428571435


In [4]:
# смещённая оценка дисперсии
vars_wages = wages_els_mmean_sqr_sum / len(wages)
print(vars_wages)

2135.074829931973


In [5]:
# несмещённая оценка дисперсии
varn_wages = wages_els_mmean_sqr_sum / (len(wages) - 1)
print(varn_wages)

2241.8285714285716


In [6]:
# среднее квадратичное отклонение (смещённоe)
print((vars_wages)**(1/2))

46.20686994302874


In [7]:
# среднее квадратичное отклонение (несмещённоe)
print((varn_wages)**(1/2))

47.34795213553139


In [8]:
# Значения с использованием библиотек:
import numpy as np
import pandas as pd

In [9]:
df = pd.DataFrame({'wages': wages})
# среднее арифметическое
print(df['wages'].mean())
# смещённая оценка дисперсии
print(df['wages'].var(ddof=0))
# несмещённая оценка дисперсии
print(df['wages'].var(ddof=1))
# среднее квадратичное отклонение (смещённоe)
print(df['wages'].std(ddof=0))
# среднее квадратичное отклонение (несмещённоe)
print(df['wages'].std(ddof=1))

73.14285714285714
2135.074829931973
2241.8285714285716
46.20686994302874
47.34795213553139


Задача 2

Для выборки из задачи 1 найти (также без использования библиотек):
1. медиану, первый и третий квартили, интерквартильное расстояние,
2. выборсы в выборке (используя для этого метод как при построении "усов" из boxplot).

Возможные неоднозначности в вычислении квантилей можно разрешать любым способом.

In [10]:
wages_sorted = sorted(wages)
print(wages_sorted)
print(len(wages))

[17, 24, 25, 30, 33, 45, 55, 57, 65, 65, 70, 75, 75, 77, 80, 84, 89, 90, 100, 150, 230]
21


In [11]:
# медиана
print(wages_sorted[int(len(wages)/2)]) # нумерация индексов с нуля, нужен (21+1)/2 = 11ый элемент, т.е., с индексом 10 

70


In [12]:
# первый квартиль
q1 = wages_sorted[int(len(wages)/4)]
print(q1)

45


In [13]:
# третий квартиль
q3 = wages_sorted[int(len(wages)*3/4)]
print(q3)

84


In [14]:
# интерквартильное расстояние - это отрезок между первым и третьим квартилями:
print([q1, q3])

[45, 84]


In [15]:
# выборсы в выборке
iqr = q3 - q1
boxplot_range_low = q1 - 1.5 * iqr 
boxplot_range_high = q3 + 1.5 * iqr
for i in range(len(wages)):
    if wages[i] < boxplot_range_low or wages[i] > boxplot_range_high:
        print(f'Выброс: сумма {wages[i]} с индексом {i}')

Выброс: сумма 230 с индексом 11
Выброс: сумма 150 с индексом 20


In [16]:
# Значения с использованием библиотек:
# медиана, первый квартиль, третий квартиль
print(df['wages'].quantile([0.25, 0.5, 0.75]))
# интерквартильное расстояние
print(np.quantile(wages, [0.25, 0.75]))

0.25    45.0
0.50    70.0
0.75    84.0
Name: wages, dtype: float64
[45. 84.]


In [17]:
# выборсы в выборке
q1 = df['wages'].quantile(0.25)
q3 = df['wages'].quantile(0.75)
iqr = q3 - q1
boxplot_range = (q1 - 1.5 * iqr, q3 + 1.5 * iqr)
outliers = df.loc[(df['wages'] < boxplot_range[0]) | (df['wages'] > boxplot_range[1])]
print(outliers)

    wages
11    230
20    150


Задача 3

В университет на факультеты A и B поступило равное количество студентов, а на факультет C студентов поступило столько же, сколько на A и B вместе. Вероятность того, что студент факультета A сдаст первую сессию, равна 0.8. Для студента факультета B эта вероятность равна 0.7, а для студента факультета C - 0.9. Студент сдал первую сессию. Какова вероятность, что он учится: a) на факультете A? б) на факультете B? в) на факультете C?

Замечание: да, эта задача не на тему описательных статистик, но тема важная, и её стоит иногда освежать в памяти.

In [18]:
# Если на факультеты A и B поступило равное количество студентов, 
# а на факультет C студентов поступило столько же, сколько на A и B,
# то на А и В учится по 1/4 всех студентов (0.25), а на С 1/2 (0/5).
# Чтобы оценить безусловную вероятность сдачи первой сессии, воспользуемся формулой полной вероятности:
P = 0.25*0.8 + 0.25*0.7 + 0.5*0.9
print(P)

0.825


In [19]:
# По формуле Байеса:
# a) вероятность, что студент учится на факультете A:
PA = 0.25*0.8/0.825
print(PA)

0.24242424242424246


In [20]:
# б) вероятность, что студент учится на факультете В:
PB = 0.25*0.7/0.825
print(PB)

0.21212121212121213


In [21]:
# в) вероятность, что студент учится на факультете С:
PC = 0.5*0.9/0.825
print(PC)

0.5454545454545455


In [22]:
print(PA+PB+PC)

1.0


Ответ: вероятность, что студент учится: a) на факультете A: 
0.24242 б) на факультете B: 0.21212 в) на факультете C: 0.54545